### PR TITLE
protocol: api to register custom schemes to standard schemes

### LINKED ISSUE #1879: Custom protocol doesn't set host, hostname and origin
Using a custom protocol handler the `window.location` object doesn't contain the `host`, `hostname` or `origin` and therefore they all share the same local storage.

Example
```js
// http://

document = {
  "origin": "http://localhost:3000",
  "hash": "",
  "search": "",
  "pathname": "/",
  "port": "3000",
  "hostname": "localhost",
  "host": "localhost:3000",
  "protocol": "http:",
  "href": "http://localhost:3000/"
}


// myprotocol://

document = {
  "origin": "myprotocol://",
  "hash": "",
  "search": "",
  "pathname": "//my.domain",
  "port": "",
  "hostname": "",
  "host": "",
  "protocol": "myprotocol:",
  "href": "myprotocol://my.domain"
}
```

This might be as it bases custom protocol handlers on the file protocol.
Is there a way to set the host to enforce separate local storages?

### PR DESCRIPTION
**Pull request by deepak1556**
Fixes #1879 

`AddAdditionalSchemes` gets called at early stage of content [intialisation](https://code.google.com/p/chromium/codesearch#chromium/src/content/app/content_main_runner.cc&sq=package:chromium&type=cs&l=703&rcl=1434281558) and standard schemes list gets locked once its done. Couldnt come up with a better approach than extracting from environment varaible `export SCHEMES=library,test,atom`. Any better option to it ?

```
> window.location
Location {ancestorOrigins: DOMStringList, href: "test://example.com/c.html"}
ancestorOrigins: DOMStringListassign: () { [native code] }
hash: ""
host: "example.com"
hostname: "example.com"
href: "test://example.com/c.html"
origin: "test://example.com"
pathname: "/c.html"
port: ""
protocol: "test:"
reload: reload() { [native code] }
replace: () { [native code] }
search: ""
toString: toString() { [native code] }
valueOf: valueOf() { [native code] }
__proto__: Location
```

### CODE REVIEW

**Comment by frozeman** [comment]
Nice!
Did you try setting local storage for different domains?

**Comment by deepak1556** [comment]
@frozeman yup works fine, once the url is parsed initially the right way, everything will work as expected.

File: atom/app/atom_content_client.cc [hunk 1]
```diff
@@ -75,6 +77,15 @@ std::string AtomContentClient::GetProduct() const {
 void AtomContentClient::AddAdditionalSchemes(
     std::vector<std::string>* standard_schemes,
     std::vector<std::string>* savable_schemes) {
+  std::string custom_schemes;
+  scoped_ptr<base::Environment> env(base::Environment::Create());
+  env->GetVar("SCHEMES", &custom_schemes);
+  if (!custom_schemes.empty()) {
+    base::StringTokenizer schemes(custom_schemes, ",");
```
Comment (zcbenz) [inline atom/app/atom_content_client.cc]: We can simply use `base::SplitString`.

Comment (deepak1556) [inline atom/app/atom_content_client.cc]: cool, is this user-friendly enough for using environment variable for fix ? Where should i document about this btw ?

**Comment by zcbenz** [comment]
Using environment variable is probably the worst user-firendly way.

> regarding adding custom schemes to standard , that does work. So  chromium has `AddAdditionalSchemes` on contentclient to let embedders register . Thought of providing a command line flag with comma searated values of custom schemes `--register-schemes="library,test,atom"` . The problem facing now is since there is a restart of the process, the second time commandline params gets lost, any thoughts on how to fix it ?

You might try appending the command line parameters in `AtomBrowserClient::AppendExtraCommandLineSwitches`.

Also we should use `--register-standard-schemes` instead of `--register-schemes` since the `"standard"` is the key of the problem.

**Comment by frozeman** [comment]
from a user perspective it would be nicer that you set the standard scheme option when you call `protocol.registerProtocol` or introduce a `protocol.registerStandardProtocol` function.

Or otherwise how would i create a bundled `.app` file, with the set flags?

**Comment by zcbenz** [comment]
@frozeman Chromium doesn't allow adding standard schemes dynamically, so it is very hard to add an API for that. For command line switches you can use `app.commandLine.appendSwitch` in your app to append them.

**Comment by steelbrain** [comment]
@zcbenz should that parameter be added before the ready event or any time before we spawn a browser window?

**Comment by frozeman** [comment]
@steelbrain yes, see here: https://github.com/atom/electron/blob/master/docs/api/chrome-command-line-switches.md

**Comment by deepak1556** [comment]
@zcbenz have provided an api on app to register. `app.registerStandardSchemes` , how does it look ?

File: atom/app/atom_content_client.cc [hunk 2]
```diff
@@ -75,6 +75,17 @@ std::string AtomContentClient::GetProduct() const {
 void AtomContentClient::AddAdditionalSchemes(
     std::vector<std::string>* standard_schemes,
     std::vector<std::string>* savable_schemes) {
+  auto command_line = base::CommandLine::ForCurrentProcess();
+  auto custom_schemes = command_line->GetSwitchValueASCII(
+      switches::kRegisterStandardSchemes);
+  if (!custom_schemes.empty()) {
+    std::vector<std::string> schemes;
+    base::SplitString(custom_schemes, ',', &schemes);
+    for (const auto& scheme : schemes) {
```
Comment (zcbenz) [inline atom/app/atom_content_client.cc]: We can use a simpler version:

```c++
standard_schemes->insert(standard_schemes.end(), custom_schemes.begin(), custom_schemes.end());
```

I don't think we need to test whether a scheme is empty.

File: atom/browser/api/atom_api_app.cc [hunk 3]
```diff
@@ -215,6 +216,15 @@ void App::SetAppUserModelId(const std::string& app_id) {
 #endif
 }
 
+void App::RegisterStandardSchemes(const std::vector<std::string>& schemes) {
```
Comment (zcbenz) [inline atom/browser/api/atom_api_app.cc]: `registerStandardSchemes` is more suitable to be a method of `protocol` module.

File: atom/browser/api/atom_api_app.cc [hunk 4]
```diff
@@ -215,6 +216,15 @@ void App::SetAppUserModelId(const std::string& app_id) {
...
+  std::stringstream stream;
+  auto vectorToString = [&stream](const std::string& scheme) {
+    stream << scheme << ",";
+  };
+  std::for_each(schemes.begin(), schemes.end(), vectorToString);
```
Comment (zcbenz) [inline atom/browser/api/atom_api_app.cc]: There is `base::JoinString`.

The string concatenating should be put in `AtomBrowserClient::SetCustomSchemes`, and `AtomBrowserClient::SetCustomSchemes` should take `vector` as parameter.

File: atom/browser/atom_browser_client.cc [hunk 5]
```diff
@@ -43,6 +43,9 @@ int kDefaultRoutingID = 2;
 // Next navigation should not restart renderer process.
 bool g_suppress_renderer_process_restart = false;
 
+// Custom schemes to be registered to standard.
+std::string custom_schemes;
```
Comment (zcbenz) [inline atom/browser/atom_browser_client.cc]: Global variable should have `g_` as prefix in its name.

File: atom/browser/atom_browser_client.cc [hunk 6]
```diff
@@ -174,6 +181,10 @@ void AtomBrowserClient::AppendExtraCommandLineSwitches(
     base::CommandLine* command_line,
     int process_id) {
   std::string process_type = command_line->GetSwitchValueASCII("type");
+
+  command_line->AppendSwitchASCII(switches::kRegisterStandardSchemes,
```
Comment (zcbenz) [inline atom/browser/atom_browser_client.cc]: I think we don't need to append `kRegisterStandardSchemes` for processes other than renderer process?

Comment (deepak1556) [inline atom/browser/atom_browser_client.cc]: Ah yes.

File: atom/app/atom_content_client.cc (hunk 2 above)
Comment (deepak1556) [inline atom/app/atom_content_client.cc]: the previous version of vectorToString would create an empty element. will be fixed.

**Comment by deepak1556** [comment]
@zcbenz fixed.

**Comment by DsXack** [comment]
:+1:

**Comment by zcbenz** [comment]
:+1:

**Comment by frozeman** [comment]
So whats the final api?

**Comment by deepak1556** [comment]
@frozeman the final api is in the `protocol` [module](https://github.com/atom/electron/blob/master/docs/api/protocol.md#protocolregisterstandardschemesvalue), `protocol.registerStandardSchemes`

**Comment by frozeman** [comment]
so its:
```js
var protocol = require('protocol');


protocol.registerStandardSchemes('my-scheme, atom, xyz');

...

app.on('ready'....)
```

?

Any docs somewhere?

**Comment by deepak1556** [comment]
```
protocol.registerStandardSchemes(['my-scheme', 'atom', 'xyz']);
```

have linked to the docs in previous comment.

**Comment by qinyang912** [comment]
after use protocol.registerStandardSchemes, in 1.0.0, the host is also empty
